ci: allow overwrite for 'latest' version

jacderida opened the pull request:
We don't want to allow explicitly versioned binaries to be overwritten, but we do need to accommodate this for the 'latest' version.

oetyng commented on sn_interface/src/messaging/system/mod.rs:
```diff
@@ -60,6 +59,18 @@ pub enum AntiEntropyKind {
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub enum SectionStateVote {
+    /// Vote to remove a node from our section
+    NodeIsOffline(NodeState),
```
There is a fundamental problem here that I mentioned to D.Rusu some months ago (so it is not something new for this PR):

`NodeIsOffline` is voted for with a `NodeState` which contains a `MembershipState`, which can be `Joined`, `Left` or `Relocated`. So we can cast a vote for `NodeIsOffline`, with a state `Joined`, which doesn't make sense?

Just to begin with I see a model mismatch, where we try to do state based and event driven handling at the same time, but never fully settling on what we are doing.
At one side, we are saying "this happened", at another side we are saying "this is our current state".

But the specific problem above I think, is us not having type safety in the system for saying `this is state x`, `this is event e` (that we saw upon that state).

grumbach replied:
Good one! I must admit I just moved and renamed that one without checking the insides of `NodeState` assuming it was correct. It's indeed a type safety issue as you say. 
Let me see what I can do to fix that. 
Since this PR is already quite big and this issue wasn't introduced by this PR, I'm thinking I could do this in a subsequent PR?

joshuef replied:
follow up PR sounds good to me

oetyng replied:
Absolutely, would not have expected you to add it here :)

oetyng commented on sn_node/src/node/messaging/handover.rs:
```diff
@@ -12,20 +12,132 @@ use crate::node::{
+        if !sap.contains_elder(&sender.name()) {
+            trace!("Ignoring invalid handover request {msg_id:?}: not DKG participant");
+            return Ok(vec![]);
+        }
```
Do we vote if we are not ourselves in it? (Does it make sense to check that we are still in the proposed sap?)

maqi replied:
This is an inner-section promotion, i.e. the existing elders shall be always stay in the new sap.
if it is not, it means that elder is dead or kicked out.

oetyng replied:
Ah, I was not looking straight. This is already decided, not voting. Not sure how I missed that.

oetyng commented on sn_node/src/node/messaging/handover.rs:
```diff
@@ -12,20 +12,132 @@ use crate::node::{
+
+        // if we have reached 2 split SAP candidates for this generation
+        // handle section split (1 to 2)
+        if let [sap1, sap2] = section_candidates_for_gen
```
I'd perhaps use the 0-based indexing, as that coincides with the rest of the references to these further down the flow. (And the prefix bit diff).

oetyng commented on sn_node/src/node/messaging/promotion.rs:
```diff
@@ -6,122 +6,135 @@
-        sap: SectionAuthorityProvider,
-        sig: SectionSig,
+        msg_id: MsgId,
+        sap1: SectionSigned<SectionAuthorityProvider>,
+        sig_share1: SectionSigShare,
+        sap2: SectionSigned<SectionAuthorityProvider>,
+        sig_share2: SectionSigShare,
```
Maybe use 0-based indexing here as well (and in rest of fn body).
(I know there was already a 1-based used before, further down.)

No biggie and can be updated later.

grumbach replied:
So I wanted to do this at first. But there's a catch: if we have sap0 and sap1 we might assume that the sap0 is the one on the 0 side of the split and sap1 on the 1 side. BUT there's nothing that enforces this! So that's why I purposely left it to sap1, sap2 so we don't have these assumptions when reading the code. There are places where it is still sap0, sap1, and it would indeed be prettier to have the same names everywhere.

oetyng commented on sn_node/src/node/mod.rs:
```diff
@@ -106,22 +106,24 @@ mod core {
         pub(crate) pending_split_sections:
             BTreeMap<Generation, BTreeSet<SectionSigned<SectionAuthorityProvider>>>,
         pub(crate) relocate_state: Option<Box<JoiningAsRelocated>>,
         // ======================== Elder only ========================
```
Just a note (not something caused by this PR): I think this comment is out of date (or perhaps unclear)? There seems to be elder only related stuff above as well?

grumbach replied:
`pending_split_sections` is indeed elder only as well 🤔

oetyng reviewed (approved):
Nice work! Great improvement, elegantly solved!

maqi commented on sn_interface/src/messaging/system/mod.rs:
```diff
@@ -114,18 +125,44 @@ pub enum NodeMsg {
+        sig_share: SectionSigShare,
+    },
+    /// After Handover consensus, the elders inform the new elder candidates that they are promoted
+    /// The candidates can aggregate the sig_shares an obtain SectionSigned proof that they are promoted
```
an => to

maqi commented on sn_interface/src/messaging/system/mod.rs:
```diff
@@ -114,18 +125,44 @@ pub enum NodeMsg {
+        /// BLS signature share of an Elder over the sap1's pubkey
+        sig_share1: SectionSigShare,
+    },
+    /// A vote about the state of a section to be aggregated in the SectionStateVote aggregator
```
maybe make this comment to follow the others' pattern by saying:
Elders votes among themselves regarding SectionStates of node_off or join_allowed

maqi commented on sn_interface/src/network_knowledge/section_authority_provider.rs:
```diff
@@ -70,6 +70,7 @@ pub struct SectionAuthorityProvider {
 }
 
 /// `SectionAuthorityProvider` candidates for handover consensus to vote on
+/// Each is signed by their own section key (the one in the SectionAuthorityProvider)
```
by their own section key => by their private section key-share

grumbach replied:
In this case the aggregation was already done at the previous step, we already have a SectionSigned piece of data (SAP) here.

maqi commented on sn_node/src/node/messaging/handover.rs:
```diff
@@ -12,20 +12,132 @@ use crate::node::{
+                self.handle_request_handover_agreement(sap, sig)
+            }
+            Ok(None) => {
+                trace!("Handover request {msg_id:?} acknowledged, waiting for more...");
```
when already aggregated, it will be returned as Ok(None) ?
if so, may worth to be careful with the log

grumbach replied:
When aggregation is finished the aggregator returns the section sig and discards the state, the next received message will start a new round of aggregation so we will indeed get a Ok(None)

grumbach replied:
With that we will indeed probably get `waiting for more...` logs when we are not anymore. What do you suggest?

maqi commented on sn_node/src/node/messaging/handover.rs:
```diff
@@ -12,20 +12,132 @@ use crate::node::{
+            .prefix()
+            .is_extension_of(&self.network_knowledge.prefix());
+        if !equal_prefix && !is_extension_prefix {
+            // Other section. We shouln't be receiving or updating a SAP for
```
shouln't => shouldn't

maqi commented on sn_node/src/node/messaging/handover.rs:
```diff
@@ -74,15 +189,65 @@ impl MyNode {
+        let (others, myself) = self.split_peers_and_self(recipients);
+        let peers = Peers::Multiple(others);
+
+        // sends a promotion message to all of the to-be-Elders with our sig_share over their new pub key
```
over their new pub key => over the new sap's public_key

maqi commented on sn_node/src/node/messaging/promotion.rs:
```diff
@@ -6,122 +6,135 @@
+
+        // Proposal from other sections shall be ignored.
+        if !our_prefix.matches(&sender.name()) {
+            trace!("Ignore promotion message {msg_id:?} from other section");
```
just a note, in case the node already got split from previous aggregated SectionSplit,
then the later on handover from old elders may got ignored here.
not an issue but may cause some confusing by the current log.
worth to log our_prefix and the sender together with the log to ease the debugging.

maqi commented on sn_node/src/node/messaging/section_state.rs:
```diff
@@ -0,0 +1,152 @@
+// Copyright 2022 MaidSafe.net limited.
```
Happy new year
and it's 2023 now  :)

grumbach replied:
#1951

oetyng commented:
bors r+

grumbach commented:
bors r+

jacderida committed:
ci: allow overwrite for 'latest' version

We don't want to allow explicitly versioned binaries to be overwritten, but we do need to
accommodate this for the 'latest' version.
